proposals/proposals/nnnn-trailing-comma-tuples-arguments-conditions.md

```diff
@@ -0,0 +1,200 @@
+
+```swift
+.frame(
+    width: 500,
+//    alignment: .leading
+) ❌ Unexpected ',' separator
+```
+
+The introduction of [parameter packs](https://github.com/apple/swift-evolution/blob/main/proposals/0393-parameter-packs.md) allows more APIs that are comma-separated lists at call site and would benefit from trailing comma.
+
+```swift
+extension [S] {
+    func sorted<each T: Comparable>(_ keyPath: repeat KeyPath<S, each T>) { }
+}
+
+arrayOfS.sorted(
+  \.a, 
+  \.b,
+//  \.c
+) ❌ Unexpected ',' separator
+```
+
+Since [#21381](https://github.com/apple/swift/pull/21381) has been merged back in 2019 **enum associated values** supports default values and are a good fit for trailing comma as well.
+
+
+**Tuples** use are very close to arguments list and, although may not be so frequently used, it seems natural that they adopt trailing comma too.
+
+**Multiple conditions** in `if`, `guard` and `while` are also comma-separated list and add, remove, reorder and comment in/out are not uncommon practice during development.
+
+```swift
+if 
+   condition1,
+   condition2,
+//   condition3
+{ ❌ Cannot convert value of type '() -> ()' to expected condition type 'Bool'
+                    
+} ❌ Expected '{' after 'if' condition
+```
+### Code Generation
+
+**Plugins** and **Macros** have made it possible to generate code using swift and trailing comma would allow generate list of arguments and conditions without worrying about a special condition for the last element.
+
+### Code Diff
+
+A tangential motivation is that trailing comma makes version-control diffs cleaner.
+
+Without trailing comma:
+```diff
+foo(
+-  a: Int
++  a: Int,
++  b: Int
+)
+```
+With trailing comma:
+```diff
+foo(
+  a: Int,
++  b: Int,
+)
+```
+
+> [!NOTE]  
+> A similar proposal was [rejected](https://forums.swift.org/t/rejected-se-0084-allow-trailing-commas-in-parameter-lists-and-tuples/2777) back in 2016 for Swift 3. It's been 8 years since that, the swift language has evolved a lot, some changes highlighted above as motivation, and the code style that "puts the terminating right parenthesis on a line following the arguments to that call" has been widely adopted by community, swift standard library codebase, swift-format, docc documentation and Xcode. Therefore, not encourage or endorse this code style doesn't hold true anymore nor is a reason for rejection.
+
+## Proposed solution
+
+This proposal adds support for trailing comma to:
+
+### Arguments and Parameters
+
+Including declaration and call of initializers, functions and enum case associated values.
+
+```swift
+
+func foo(
+    a: Int = 0, 
+    b: Int = 0, 
+) {
+}
+
+foo(
+    a: 1,
+    b: 2,
+)
+
+```
+
+### Tuples
+
+```swift
+(1, 2, 3,)
+```
+
+### Conditions
+
+Including `if`, `guard` and `while`.
+
+```swift
+if 
+   condition1,
+   condition2,
+   condition3,
+{
+                    
+}
+```
+
+## Source compatibility
+
+This change is purely additive and has no impact on existing code.
```

Comment:
This, or the above section, might be a good place to include your findings regarding the implications for parsing condition lists that you've mentioned in the pitch thread. Even if how they're parsed is formally never ambiguous to the compiler, if there are scenarios in which a reader could be confused/misled due to how trailing commas are parsed in condition lists, it'd be good to record for posterity what the intended behavior is.